make/devices.cmake

```diff
@@ -0,0 +1,754 @@
+    )
+
+    make_file_target(FILE ${OUT_RRXML_REAL_FILENAME} GENERATED)
+
+    set_target_properties(
+      ${DEFINE_DEVICE_DEVICE}
+      PROPERTIES
+        ${PACKAGE}_OUT_RRXML_REAL
+        ${CMAKE_CURRENT_SOURCE_DIR}/${OUT_RRXML_REAL_FILENAME}
+    )
+
+    add_custom_target(
+      ${DEFINE_DEVICE_ARCH}_${DEFINE_DEVICE_DEVICE}_${PACKAGE}_rrxml_real
+      DEPENDS ${OUT_RRXML_REAL}
+    )
+  endforeach()
+endfunction()
+
+function(DEFINE_BOARD)
+  # ~~~
+  # DEFINE_BOARD(
+  #   BOARD <board>
+  #   DEVICE <device>
+  #   PACKAGE <package>
+  #   PROG_TOOL <prog_tool>
+  #   [PROG_CMD <command to use PROG_TOOL>
+  #   )
+  # ~~~
+  #
+  # Defines a target board for a project.  The listed device and package must
+  # have been defined using DEFINE_DEVICE.
+  #
+  # PROG_TOOL should be an executable that will program a bitstream to the
+  # specified board. PROG_CMD is an optional command string.  If PROG_CMD is not
+  # provided, PROG_CMD will simply be ${PROG_TOOL}.
+  #
+  set(options)
+  set(oneValueArgs BOARD DEVICE PACKAGE PROG_TOOL PROG_CMD)
+  set(multiValueArgs)
+  cmake_parse_arguments(
+    DEFINE_BOARD
+    "${options}"
+    "${oneValueArgs}"
+    "${multiValueArgs}"
+    ${ARGN}
+  )
+
+  add_custom_target(${DEFINE_BOARD_BOARD})
+  foreach(ARG DEVICE PACKAGE PROG_TOOL PROG_CMD)
+    set_target_properties(
+      ${DEFINE_BOARD_BOARD}
+      PROPERTIES ${ARG} "${DEFINE_BOARD_${ARG}}"
+    )
+  endforeach()
+endfunction()
+
+function(ADD_OUTPUT_TO_FPGA_TARGET name property file)
+  make_file_target(FILE ${file} GENERATED)
+  set_target_properties(${name} PROPERTIES ${property} ${file})
+endfunction()
+
+function(ADD_FPGA_TARGET)
+  # ~~~
+  # ADD_FPGA_TARGET(
+  #   NAME <name>
+  #   TOP <top>
+  #   BOARD <board>
+  #   SOURCES <source list>
+  #   TESTBENCH_SOURCES <testbench source list>
+  #   [INPUT_IO_FILE <input_io_file>]
+  #   [EXPLICIT_MAKE_FILE_TARGET]
+  #   )
+  # ~~~
+  #
+  # ADD_FPGA_TARGET defines a FPGA build targetting a specific board.  By
+  # default input files (SOURCES, TESTBENCH_SOURCES, INPUT_IO_FILE) will be
+  # implicitly passed to MAKE_FILE_TARGET.  If EXPLICIT_MAKE_FILE_TARGET is
+  # supplied, this behavior is supressed.
+  #
+  # The SOURCES file list will be used to synthesize the FPGA images.
+  # INPUT_IO_FILE is required to define an io map. TESTBENCH_SOURCES will be
+  # used to run test benches.
+  #
+  # Targets generated:
+  #
+  # * <name>_eblif - Generate eblif file.
+  # * <name>_synth - Alias of <name>_eblif.
+  # * <name>_route - Generate place and routing synthesized design.
+  # * <name>_bit - Generate output bitstream.
+  #
+  # Outputs for this target will all be located in
+  # ~~~
+  # ${CMAKE_CURRENT_BINARY_DIR}/${NAME}/${ARCH}-${DEVICE_TYPE}-${DEVICE}-${PACKAGE}
+  # ~~~
+  #
+  # Output files:
+  #
+  # * ${TOP}.eblif - Synthesized design
+  # * ${TOP}_io.place - IO placement.
+  # * ${TOP}.route - Place and routed design
+  # * ${TOP}.hlc - Place and routed design
```

Comment:
HLC files are specific to iCE40. It is likely that `.fasm` files will be the generic output in the future.

Might be worth linking to the VPR documentation for each file type?

Reply:
I linked to eblif and .route format links in VPR